handlers/auth.go

```diff
@@ -364,179 +365,148 @@ func Logout(w http.ResponseWriter, r *http.Request) {
 	data.Meta.NoTrack = true
 
 	if handleTemplateError(w, r, "auth.go", "ResetPassword", "", resetPasswordTemplate.ExecuteTemplate(w, "layout", data)) != nil {
 		return // an error has occurred and was processed
 	}
 }
 
-// ResetPasswordPost resets the password to the value provided in the form, given that the user is authenticated.
+// ResetPasswordPost resets the password to the value provided in the form, given that the hash is correct.
 func ResetPasswordPost(w http.ResponseWriter, r *http.Request) {
-	logger := logger.WithField("route", r.URL.String())
+	hash := r.FormValue("hash")
+	errFields := map[string]interface{}{
+		"route": r.URL.String(),
+		"hash":  hash,
+	}
 
-	user, session, err := getUserSession(r)
+	// fetch user from db by hash
+	user := struct {
+		ID      uint64    `db:"id"`
+		ResetTs time.Time `db:"password_reset_ts"`
+	}{}
+	err := db.FrontendWriterDB.Get(&user, `
+			SELECT users.id, password_reset_ts
+			FROM users 
+			WHERE password_reset_hash = $1`, hash)
 	if err != nil {
-		logger.Errorf("error retrieving session: %v", err)
-		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		if errors.Is(err, sql.ErrNoRows) {
+			utils.SetFlash(w, r, authSessionName, "Error: Invalid or outdated reset link, please retry.")
+		} else {
+			utils.LogError(err, "error resetting password", 0, errFields)
+			utils.SetFlash(w, r, authSessionName, authInternalServerErrorFlashMsg)
+		}
+		http.Redirect(w, r, "/requestReset", http.StatusSeeOther)
 		return
 	}
 
-	if !user.Authenticated {
-		session.AddFlash("Error: You are not authenticated (or did not use the correct reset-link).")
-		session.Save(r, w)
-		http.Redirect(w, r, "/confirmation", http.StatusSeeOther)
-		return
-	}
+	errFields["user_id"] = user.ID
 
-	err = r.ParseForm()
-	if err != nil {
-		utils.LogError(err, "error parsing form", 0)
-		session.AddFlash(authInternalServerErrorFlashMsg)
-		session.Save(r, w)
-		http.Redirect(w, r, "/confirmation", http.StatusSeeOther)
+	// check expired ts
+	if user.ResetTs.Add(authEmailExpireTime).Before(time.Now()) {
+		utils.SetFlash(w, r, authSessionName, "Error: Invalid or outdated reset link, please retry.")
+		http.Redirect(w, r, "/requestReset", http.StatusSeeOther)
 		return
 	}
 
+	// update password
 	pwd := r.FormValue("password")
 	pHash, err := bcrypt.GenerateFromPassword([]byte(pwd), 10)
 	if err != nil {
-		logger.Errorf("error generating hash for password: %v", err)
-		session.AddFlash(authInternalServerErrorFlashMsg)
-		session.Save(r, w)
-		http.Redirect(w, r, "/confirmation", http.StatusSeeOther)
+		utils.LogError(err, "error generating hash for password", 0, errFields)
+		utils.SetFlash(w, r, authSessionName, authInternalServerErrorFlashMsg)
+		http.Redirect(w, r, "/requestReset", http.StatusSeeOther)
 		return
 	}
 
-	err = db.UpdatePassword(user.UserID, pHash)
+	_, err = db.FrontendWriterDB.Exec("UPDATE users SET password = $1, password_reset_hash = NULL WHERE id = $2", pHash, user.ID)
```

Comment:
**question:** I don't see a point to reset the password reset hash in the "New Behavior" list but I assume everything is as intended as it makes no sense to me to keep it saved after the password has been changed?

Reply:
The reset token isn't supposed to work anymore after it's used once, correct. This is standard practice.